File: smart_home_app/smart_home_app/lib/logic/services/device_service.dart
import 'package:smart_home_core/models.dart';

import '../../models/generic_device.dart';
import 'interfaces/device_service.dart';

class DeviceService implements IDeviceService {
  @override
  Future<List<ScannedDevice>> getDevices() async {
    return [
      ScannedDevice(
        type: DeviceType.airQuality,
        name: 'Kvaliteta zraka',
        hostname: 'air-quality-station.local',
        uuid: '',
      ),
    ];

    /*var username = AppContext.instance.currentUser.username;

    var client = BackendClient();
    var response = await client.httpGet('/api/users/$username/devices');

    var devices = _parseDevices(response);
    devices.sort((a, b) => a.name.compareTo(b.name));

    return devices;*/
  }

  @override
  Future<List<ScannedDevice>> getDevicesFromCache() async {
    return [];
    /*var file = _getDevicesFile();
    if ((await file.exists()) == false) {
      return [];
    }

    var data = await file.readAsString();
    var json = jsonDecode(data);
    var devices = _parseDevices(json);

    return devices;*/
  }

  @override
  Future<void> saveDevicesToCache(List<ScannedDevice> devices) async {
    /*var file = _getDevicesFile();
    var data = devices.map((e) => e.toJson()).toList();
    var json = jsonEncode(data);

    await file.writeAsString(json);*/
  }

  @override
  Future<void> deleteDevicesFromCache() async {
    /*var file = _getDevicesFile();
    if ((await file.exists())) {
      await file.delete();
    }*/
  }

  /*
  File _getDevicesFile() {
    var appDir = AppContext.instance.appDirectory;
    return File(join(appDir, 'devices.json'));
  }

  List<ScannedDevice> _parseDevices(List<dynamic> json) {
    var devices = <ScannedDevice>[];
    for (var data in json) {
      try {
        var device = ScannedDevice.fromJson(data);
        devices.add(device);
      } catch (e) {
        log(data.toString() + e.toString());
      }
    }
    return devices;
  }*/
}

File: smart_home_angular/src/app/core/services/device-service.ts
import { Injectable } from '@angular/core';
import { Device } from '../../core/models/device';
import { DeviceTypeHelper } from '../../core/models/device_type';

@Injectable({
  providedIn: 'root',
})
export class DeviceService {
  public async getDevices(): Promise<Device[]> {
    let json = localStorage.getItem('devices');
    if (json == null) return <Device[]>[];

    return this.parseDevices(json);
  }

  private parseDevices(text: string) {
    let json = JSON.parse(text);
    let devices: Device[] = json.map((e: any) => this.parseDevice(e));
    return devices;
  }

  private parseDevice(data: any) {
    let id = data['id'] as string;
    let name = data['name'] as string;
    let type = DeviceTypeHelper.parseType(data['type'] as string);
    let hostname = data['hostname'] as string;
    return new Device(id, name, type, hostname);
  }
}
